### ISSUE TITLE
Segmentation fault

### ISSUE THREAD

**Issue by rittman** [colearendt/xlsx#83]
Hi,

in R version 3.4.0 (2017-04-21) this package is causing a segmentation fault on loading in 64-bit ubuntu.

Regards,

Tim

**Comment by msperlin** [comment]
Same problem here.. I just came to report the issue and saw the message. 

I tried reinstalling all R packages. Nothing worked.

Here's the issue:

> install.packages('xlsx')
Installing package into ‘/home/msperlin/R/x86_64-pc-linux-gnu-library/3.4’
(as ‘lib’ is unspecified)
trying URL 'https://cran.rstudio.com/src/contrib/xlsx_0.5.7.tar.gz'
Content type 'application/x-gzip' length 312839 bytes (305 KB)
==================================================
downloaded 305 KB

* installing *source* package ‘xlsx’ ...
** package ‘xlsx’ successfully unpacked and MD5 sums checked
** R
** inst
** preparing package for lazy loading
** help
*** installing help indices
** building package indices
** installing vignettes
** testing if installed package can be loaded
Segmentation fault
ERROR: loading failed
* removing ‘/home/msperlin/R/x86_64-pc-linux-gnu-library/3.4/xlsx’
Warning in install.packages :
  installation of package ‘xlsx’ had non-zero exit status

The downloaded source packages are in
	‘/tmp/Rtmpd0tmyz/downloaded_packages’


Here my details:

R version 3.4.0 (2017-04-21)
Platform: x86_64-pc-linux-gnu (64-bit)
Running under: Linux Mint 18.1

Matrix products: default
BLAS: /usr/lib/libblas/libblas.so.3.6.0
LAPACK: /usr/lib/lapack/liblapack.so.3.6.0

locale:
 [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C               LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8     LC_MONETARY=pt_BR.UTF-8   
 [6] LC_MESSAGES=en_US.UTF-8    LC_PAPER=pt_BR.UTF-8       LC_NAME=C                  LC_ADDRESS=C               LC_TELEPHONE=C            
[11] LC_MEASUREMENT=pt_BR.UTF-8 LC_IDENTIFICATION=C       

attached base packages:
[1] stats     graphics  grDevices utils     datasets  methods   base     

loaded via a namespace (and not attached):
[1] compiler_3.4.0 tools_3.4.0

**Comment by colearendt** [comment]
I will see if there is a way to reproduce this on Travis.  In any case, would either/both of you mind taking a look at the following:
- What is the value of the JAVA_HOME environment variable?  Do you have a jdk installed?  Does `rJava` install/load without issue?  `xlsxjars`?
- Have you tried `sudo R CMD javareconf` per the "Installation" heading [here](https://www.rforge.net/rJava/)?  This fixes many issues.
- Does `devtools::install_github('dragua/xlsx')` cause the same problem?  How about `devtools::install_github('dragua/xlsx',ref='41d865b')`?
- One last question.  Do you have `java.parameters` set under options? `getOption('java.parameters')`

I am hopeful that the answer to these questions will provide a bit of further direction.  Interesting timing that both of you had the same issue within a few hours of one another!

**Comment by PJones7** [comment]
Hi,
We are also experiencing the same problem (It has previously been working), we are running a slightly older version of R "3.3.2".
We are able to load rjava and xlsxjars without any issues.
We did receive security updates last night, and upon rebooting the system we are experiencing the same problem as above, this is also within the timeline of above.

sudo R CMD javareconf - Yes no luck with this.

getOption('java.parameters')
[1] "-Xmx512m"

Distributor ID: Ubuntu
Description:    Ubuntu 16.04.1 LTS
Release:        16.04
Codename:       xenial

**Comment by msperlin** [comment]
Hi Cole, 

I was able to reproduce it in my office computer. Exactly the same problem. Not sure if is related but I had to mess around with pandoc for a book compilation. 

Thanks for the quick reply. Here are my answers:

**What is the value of the JAVA_HOME environment variable? Do you have a jdk installed? Does rJava install/load without issue? xlsxjars?**

Yes, both install fine in R. I've been using xlsx for a long time. All requirements are there.

**Have you tried sudo R CMD javareconf per the "Installation" heading here?** 

Yes, it worked fine. Tried installing it again, same error.

**Does devtools::install_github('dragua/xlsx') cause the same problem? How about devtools::install_github('dragua/xlsx',ref='41d865b')?**

Same problem in both.

**One last question. Do you have java.parameters set under options? getOption('java.parameters')**

No, its NULL.

**Comment by colearendt** [comment]
Very interesting.  I know little about seg faults, but am expecting something in the lower level Java code.  Wondering also if the kernel security updates mentioned are contributing.  That would make sense for the timing!

Do you guys mind trying to install commit 7ed97ff with `devtools::install_github('dragua/xlsx', ref='7ed97ff')`?

I disabled onLoad and onAttach to see if they contribute.  If the install doesn't work, try from my fork - But I think you should be able to reference that commit from the main repo

**Comment by atclaus** [comment]
That latest dev works for me right now:

    > library(devtools)
    > devtools::install_github('dragua/xlsx', ref='7ed97ff')
    Downloading GitHub repo dragua/xlsx@7ed97ff
    from URL https://api.github.com/repos/dragua/xlsx/zipball/7ed97ff
    Installing xlsx
    '/usr/lib/R/bin/R' --no-site-file --no-environ --no-save --no-restore --quiet  \
      CMD INSTALL '/tmp/Rtmpu5vMUL/devtools235e49fbb629/dragua-xlsx-7ed97ff'  \
      --library='/usr/local/lib/R/site-library' --install-tests

    * installing *source* package ‘xlsx’ ...
    ** R
    ** inst
    ** tests
    ** preparing package for lazy loading
    ** help
    *** installing help indices
    ** building package indices
    ** installing vignettes
    ** testing if installed package can be loaded
    * DONE (xlsx)
    > library(xlsx)
    Loading required package: rJava
    Loading required package: xlsxjars

**Comment by colearendt** [comment]
Interesting.  @atclaus, do you mind working with me a bit more, as this is likely not resolved.  

After installing, can you execute the following and see if you get the segfault (and at which step)
```r
library('xlsx')

.jpackage('xlsx')

wb <- createWorkbook()
rm(wb) 

 # what's your java  version?  Need > 1.5.0.
 jversion <- .jcall('java.lang.System','S','getProperty','java.version')
 if (jversion < "1.5.0")
  stop(paste("Your java version is ", jversion,".  Need 1.5.0 or higher.", sep=""))
```

**Comment by colearendt** [comment]
Thanks to everyone's help working on this.  I have been able to recreate the issue locally, and it looks very much like the issue is due to the security updates released for Ubuntu, and seems to affect the `rJava` package - particularly when `.jpackage` is called within `.onLoad` or `.onAttach` during install (the canonical approach).  If the package has already been installed, then `.onLoad` and `.onAttach` have no problems. 

In any case, feel free to follow and contribute to the [underlying issue](https://github.com/s-u/rJava/issues/110).  Meanwhile, I have a workaround that can be installed with
```r
devtools::install_github('colearendt/xlsx',ref='bug/segfault')
```
The workaround bypasses the usual `.onLoad`/`.onAttach` if `grepl('linux-gnu',R.version$os)` and prints an informative package startup message.  In order to use the package, the user must then call the following to load the necessary `.jar` files.
```r
rJava::.jpackage('xlsxjars')
rJava::.jpackage('xlsx')
```

This is admittedly an ugly workaround, and I will try to gauge the speed of the fix to see whether a more lasting workaround is justifiable.

For those who have had the issue, please test to be sure whether the workaround is successful for you.

**Comment by msperlin** [comment]
Thanks colearendt, the fix worked!

**Comment by atclaus** [comment]
So when I relaunched R as normal user just now, it fails to load rJava. Launching `sudo R` and I can load `library(xlsx)`. My server automatically updates and reboots every night so wonder if that changed things.

    > library(xlsx)
    Loading required package: rJava
    Error: package or namespace load failed for ‘rJava’:
     .onLoad failed in loadNamespace() for 'rJava', details:
      call: dyn.load(file, DLLpath = DLLpath, ...)
      error: unable to load shared object '/usr/local/lib/R/site-library/rJava/libs/rJava.so':
      libjvm.so: cannot open shared object file: No such file or directory
    Error: package ‘rJava’ could not be loaded

I tried running `R CMD javareconf` and `sudo R CMD javareconf` but that didn't resolve the issue this time (did when I first installed rJava). Had to `unset JAVA_HOME`.

Using `7ed97ff` I get the following in `sudo R`

    > library('xlsx')
    Loading required package: rJava
    Loading required package: xlsxjars
    > .jpackage('xlsx')
    > wb <- createWorkbook()
    Error in .jnew("org/apache/poi/xssf/usermodel/XSSFWorkbook") :
      java.lang.ClassNotFoundException
    > rm(wb)
    Warning message:
    In rm(wb) : object 'wb' not found
    > jversion <- .jcall('java.lang.System','S','getProperty','java.version')
    >  if (jversion < "1.5.0")
    +   stop(paste("Your java version is ", jversion,".  Need 1.5.0 or higher.", sep=""))
    > jversion
    [1] "1.8.0_131"

but in regular `R`:

    > .jpackage('xlsx')
    Segmentation fault (core dumped)

Tried the segfault ref. Installed under `sudo R`. Installed and then prompted, both manual lines worked and then it loaded. 

    * DONE (xlsx)
    > library(xlsx)
    Loading required package: rJava
    Loading required package: xlsxjars
    .jpackage is currently failing on Linux-GNU.  Best to run it manually for now with rJava::.jpackage('xlsxjars')
    .jpackage is currently failing on Linux-GNU.  Best to run it manually for now with rJava::.jpackage('xlsx')
    > rJava::.jpackage('xlsxjars')
    > rJava::.jpackage('xlsx')
    > library(xlsx)

But I cannot run in normal `R` or if I close the instance of `sudo R` and relaunch it.

    > rJava::.jpackage('xlsxjars')
    Segmentation fault (core dumped)

and

    > rJava::.jpackage('xlsx')
    Segmentation fault (core dumped)

**Comment by colearendt** [comment]
Very interesting.  Comforting that the workaround worked at first, at least.  I expect the errors mentioned from `7ed97ff`, as `.jpackage('xlsx')` and `.jpackage('xlsxjars')` is required to add the necessary `.jar` files to the classpath (and that commit stopped calling them from `.onAttach`/`.onLoad`).

You are right that a reloaded R session does not succeed with the workaround for me either... very unfortunate, and quite unexpected behavior.  It's like the install is required each time to keep the memory allocation valid.  From the linked ticket, `.jinit()` seems to be the problem (and `.jpackage()` calls `.jinit` before adding all the .jar files from the package to the classpath).

The Ubuntu security update was related to heap space security, so it seems that whatever heap space `.jinit` is trying to use for the jvm is not allowed.

**Comment by msperlin** [comment]
Can't we just rollback the linux update? 

I apologize if this is a dumb question, I'm using linux exclusivelly for just a couple of months, and a bit frustated that my R code was broken so easilly. Is this type of problem normal with Linux?

**Comment by colearendt** [comment]
I would definitely expect rolling back the linux update to fix the issue, as I did not have the issue until updating (I updated specially to test whether it was the cause, as I am normally slow to update software).  

There are definitely ways to roll-back, although I am not an expert: https://askubuntu.com/questions/610328/roll-back-kernel

I definitely think the pros of linux outweigh issues such as these, but that's just my opinion - so I would encourage you not to lose hope.  From the referenced question, it looks like kernel updates have potential for this sort of issue, but it is probably advisable to update in a way that makes roll-back easier.  Software updates on any system are something to be aware of - update on my family's windows computer messed up the wifi driver, and I have lost count of the number of cell phones that had crippling issues after software updates.

**Comment by larry77** [comment]
Hello,
Just for the record, same issue on Debian stable (freshly released) + R 3.4. This is very annoying as rJava does install, but a number of other packages depending on it do not (xlsx, RJSDMX, openNLP etc...). The suggested workaround does not work on my system (neither as normal user nor as a superuser). Rather than thinking about downgrading your system/using an older kernel (BTW, I went through pains to install a 3.x kernel on my machine with no avail), is there a way to fix this at the rJava level? Should we get in touch with the developer of rJava? What do you think? This is really annoying for me as it impacts some crucial packages for my work.

**Comment by colearendt** [comment]
I definitely agree at the level of irritation - we should be able to fix at the rJava level.  Please see the [related issue on rJava](https://github.com/s-u/rJava/issues/110) - I have narrowed down to what I believe is the C function that is having trouble.  I am not a C developer, though - so I am not able to go any further.  I also have not found any really consistent workaround on my end either - as I mentioned, the proposed workaround seems to work only on install, and not in subsequent R sessions (or at least not consistently)

**Comment by larry77** [comment]
Thanks for your help, no criticism at all on my side, just some frustration because this is impacting some of my "pet" R packages. I am an R power user, but do not ask me to write in C (!).

**Comment by colearendt** [comment]
No worries :) I definitely understand the frustration!  One other thing - I think I would say that though rJava can _install_, it cannot _work_, which is where the other packages have issue (because they presume upon rJava working and initialize on install / load / attach).  At least on my computer, the following fails with a segfault
```r
library(rJava)
.jinit()
```

**Comment by larry77** [comment]
Same situation on my machine. You are right: there is a bug in rJava, which is the root of all evil. It affects anything it depends on it, but because it is a "silent" bug, you are not aware of it until you try to use some java dependent package in R.

**Comment by msperlin** [comment]
I just rolled back to the previous kernel and it all worked fine. It is also very easy, just install boot-customizer and select the previous build and restart.

Thanks for the reply colearendt. I'm definitely now giving up on Linux so easily. Appreciate all the efforts in solving this issue.  This is what makes the R community so great.

**Comment by colearendt** [comment]
Per the related `rJava` issue, looks like a potential fix to the kernel - haven't tested yet:

> I have updated my kernel to the v4.12-rc7/	2017-06-26 04:00 version following the following steps suggested by WinEunuuchs2Unix on this thread https://askubuntu.com/questions/119080/how-to-update-kernel-to-the-latest-mainline-version-without-any-distro-upgrade

Better link: https://askubuntu.com/a/885165

**Comment by larry77** [comment]
Hello,
I am on debian stable (stretch). Yesterday I received a kernel update and all my problems with rJava, xlsx, RJSDMX etc...have disappeared. This is just to say that the fix on any debian derivative does not necessarily mean a roll back of the kernel plus eventually uninstalling a security patch (quite scary stuff, if you think about it), but also moving forward and keeping your system up-to-date.

**Comment by colearendt** [comment]
Just updated Ubuntu kernel to : `4.4.0-83-generic`

From my testing, it looks like this issue is resolved (the issue was with `4.4.0-81-generic` for me).  Required a `sudo R CMD javareconf` for me, but everything is working fine now.  Please reopen if any issues - it looks like the newest released kernels solve the problem.

A helpful command to search for kernel images available, on Ubuntu at least:

```
sudo apt-cache search linux-image
```

**Comment by atclaus** [comment]
Working for me now (on `4.4.0-83-generic`).

Initially it only worked within `sudo R`. In addition to `sudo R CMD javareconf`, I (had to) run `R CMD javareconf` and `unset JAVA_HOME`
